firmware/src/main.cpp

```diff
@@ -98,6 +105,76 @@ void quaternionToEulerGI(sh2_GyroIntegratedRV_t *rotational_vector, euler_t *ypr
 	quaternionToEuler(rotational_vector->real, rotational_vector->i, rotational_vector->j, rotational_vector->k, ypr, degrees);
 }
 
+void quaternionToControlAngles(sh2_Quaternion_t *quaternion, euler_t *ypr)
+/*
+The extraction of the YPR angles does not correspond to any of the standard Euler angles sequences.
+The reason for this is to provide the best possible feeling to the pilot, preventing the hand from
+tilting too far in the corners and ensuring symmetry for the pitch and roll angles.
+*/
+{
+	// ypr->pitch = atan2((2 * (quaternion->x * quaternion->y + quaternion->w * quaternion->z)), (quaternion->w*quaternion->w + quaternion->x*quaternion->x - quaternion->y*quaternion->y - quaternion->z*quaternion->z));
+	// ypr->roll = atan2((- 2 * (quaternion->x * quaternion->z - quaternion->w * quaternion->y)), (quaternion->w*quaternion->w + quaternion->x*quaternion->x - quaternion->y*quaternion->y - quaternion->z*quaternion->z));
+
+	// float torsion_quat_norm = sqrt(quaternion->w*quaternion->w + quaternion->x*quaternion->x);
+	// sh2_Quaternion_t torsion_quaternion = {quaternion->w/torsion_quat_norm, quaternion->x/torsion_quat_norm, 0, 0};
+	// ypr->yaw =
+
+    //yaw rotation around x
+    //pitch rotation around z
+    //roll rotation around y
+	// The quaternion order and sign has to be changed to ensure that the YPR values are correct
+    float q0 = quaternion->w; //w
+    float q1 = quaternion->z; //x
+    float q2 = - quaternion->x; //y
+    float q3 = quaternion->y; //z
```

Comment:
What quaternion convention is the input in?
https://www.iri.upc.edu/people/jsola/JoanSola/objectes/notes/kinematics.pdf Chapter 3
e.g. Eigen PX4 and so on use Hamilton convention.

Reply:
Happy to adopt any standard here! Open to your suggestions. We should check what standard comes out of the BNO086 from Bosch and convert it to whatever standard we will use for the remote.

I've been using [this script](https://github.com/thecountoftuscany/PyTeapot-Quaternion-Euler-cube-rotation) which uses another standard too, that's why it's a bit of a mess currently.